bigquery/google/cloud/bigquery/dataset.py

```diff
@@ -147,6 +147,45 @@ def table(self, table_id):
         """
         return TableReference(self, table_id)
 
+    @classmethod
+    def from_api_repr(cls, resource):
+        project = resource['projectId']
+        dataset_id = resource['datasetId']
+        return cls(project, dataset_id)
+
+    def to_api_repr(self):
+        return {
+            'projectId': self._project,
+            'datasetId': self._dataset_id,
+        }
+
+    def _key(self):
+        """A tuple key that unique-ly describes this field.
+
+        Used to compute this instance's hashcode and evaluate equality.
+
+        Returns:
+            tuple: The contents of this :class:`DatasetReference`.
+        """
+        return (
+            self._project,
+            self._dataset_id,
+        )
+
+    def __eq__(self, other):
+        if not isinstance(other, DatasetReference):
+            return NotImplemented
```

Comment:
s/return/raise/?

Reply:
And I would say it's TypeError or something like that. But maybe this is a convention?

Reply:
Yeah, should be raise or maybe just return False actually. Copy-pasted from SchemaField, so I'll update that too.

Reply:
`return NotImplemented` is [idiomatic Python](https://docs.python.org/2/library/constants.html#NotImplemented) for a binary operator implementation:  it allows the "right-hand-side" version to be invoked, in case the `other` class *does* know how to do the comparison.